src/Compilers/CSharp/Portable/BoundTree/BoundBinaryOperator.UncommonData.cs

```diff
@@ -0,0 +1,42 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class BoundBinaryOperator
+    {
+        internal class UncommonData
+        {
+            public static UncommonData? CreateIfNeeded(ConstantValue? constantValue, MethodSymbol? method, TypeSymbol? constrainedToType, ImmutableArray<MethodSymbol> originalUserDefinedOperatorsOpt)
+            {
+                if (constantValue != null || method is not null || constrainedToType is not null || !originalUserDefinedOperatorsOpt.IsDefault)
+                {
+                    return new UncommonData(constantValue, method, constrainedToType, originalUserDefinedOperatorsOpt);
+                }
+
+                return null;
+            }
+
+            public readonly ConstantValue? ConstantValue;
```

Comment:
I could be convinced that `ConstantValue` is common enough to not need to be put in here.

Reply:
I would want to go based off some kind of measurement for this, i.e. if we compile Roslyn with some additional instrumentation on this type, can we count how many binary operators have this versus how many don't.

Reply:
Obviously this would vary a lot based on what you're compiling. So I would expect projects with lots of binary operators on constant values to be generated code for example.